Refactor BabyRegistrationForm

AlexNaunt5 opened the pull request:
Refactor the BabyRegistrationForm view to follow MVVM architecture and introduces several foundational improvements to align with SOLID principles.

# Changes Made

- Migrated form logic and validation to a dedicated BabyRegistrationViewModel
- Introduced Baby model struct with strongly typed enums for:
1. Gender
2. WeightUnit
3. HeightUnit
- Replaced hardcoded strings with enum values where applicable
- Decoupled form state management from the view layer
- Improved input validation logic (first/last name, height, weight, blood type)

# Why This Refactor?
- Improves testability of form logic and validation
- Sets up a clear separation between UI, business logic, and models
- Prepares the codebase for:
1. Dependency injection (e.g., service layer for backend integration)
2. String localization
3. Reusable UI components

# Next Steps
- Extract hardcoded strings for localization
- Create reusable components for common fields (e.g., LabeledTextField, LabeledPicker)
- Add unit tests for BabyRegistrationViewModel
- Integrate with backend once available

AlexNaunt5 committed:
initial version baby registration form

AlexNaunt5 committed:
add text alert for lack of answer on a specific field and vertification of characters on each field

AlexNaunt5 committed:
Merge branch 'develop' into feature/BabyRegistrationForm

AlexNaunt5 committed:
Refactor BabyRegistrationForm to MVVM architecture with enums and basic validation logic moved to ViewModel

AlexNaunt5 committed:
formatting

AlexNaunt5 committed:
convert baby registration form into modular UI

AlexNaunt5 commented:
Complete refactor of the BabyRegistrationForm to use reusable UI components provided by the DesignKit module. This aligns with our modular architecture, where each screen will use shared components for consistency and maintainability.

### Changes

**Created reusable components in DesignKit:**
- `DKTextField` – Customizable text field with styling and keyboard options.
- `DKDatePicker` – Simplified date picker with a title.
- `DKPicker` – Generic picker component supporting any `Identifiable & CustomStringConvertible` type.
- `DKButton` – Styled primary button component.

**Updated `BabyRegistrationForm` to use:**
- `DKTextField` for input fields like name, height, weight, etc.
- `DKDatePicker` for selecting the date of birth.
- `DKPicker` for gender, weight unit, and height unit selection.
- `DKButton` for form submission.

**Added `CustomStringConvertible` conformance** to:
- `Gender`
- `WeightUnit`
- `HeightUnit`

This enables proper display support inside the generic `DKPicker`.

AlexNaunt5 committed:
crashlytics/analytics configuration and based implementation

IsaacFrancoNino commented on BabbleBuddyApp/BabbleBuddyApp.xcodeproj/project.pbxproj:
```diff
@@ -174,6 +174,7 @@
 				INFOPLIST_FILE = BabbleBuddyApp/Info.plist;
 				INFOPLIST_KEY_UIApplicationSupportsIndirectInputEvents = YES;
 				INFOPLIST_KEY_UILaunchStoryboardName = LaunchScreen;
+				INFOPLIST_KEY_UIMainStoryboardFile = Main;
```
This file still has a reference to the Main Storyboard, which is not used in this project, also the app will not run with it.
Could you remove the reference or remove the file?

IsaacFrancoNino commented on DesignKit/DesignKit/Baby.swift:
```diff
@@ -0,0 +1,40 @@
+    var lastName: String
+    var dateOfBirth: Date
+    var gender: Gender
+    var weight: String
```
Do weight and height need to be String types here, or could they be Double or Float instead?

IsaacFrancoNino commented on DesignKit/DesignKit/BabyRegistrationForm.swift:
```diff
@@ -0,0 +1,55 @@
+import SwiftUI
+
+struct BabyRegistrationForm: View {
+    @StateObject private var viewModel = BabyRegistrationViewModel()
```
Your instance of the viewModel is hardcoded, you could pass the viewModel as a parameter.

AlexNaunt5 replied:
Solved

IsaacFrancoNino commented on DesignKit/DesignKit/BabyRegistrationForm.swift:
```diff
@@ -0,0 +1,55 @@
+
+    var body: some View {
+        Form {
+            DKTextField("First Name", text: $viewModel.baby.firstName)
```
You could create a Localizable.strings file to store all your module strings and reference them here to avoid hardcoding.

IsaacFrancoNino commented on DesignKit/DesignKit/BabyRegistrationViewModel.swift:
```diff
@@ -0,0 +1,47 @@
+        let heightWeightRegex = "^[0-9.'\\\"]+$"
+
+        if baby.firstName.trimmingCharacters(in: .whitespaces).isEmpty || !NSPredicate(format: "SELF MATCHES %@", nameRegex).evaluate(with: baby.firstName) {
+            alertMessage = "Please enter a valid first name using only letters."
```
Same here with the Localizable.strings file.

IsaacFrancoNino reviewed (changes requested):
Left some comments.
While your `Components` folder is in the correct module, the `BabyRegistrationForm` and `BabyRegistrationViewModel` should be moved to their own module since each screen should live in a separate module.
Also, consider moving the `Baby` model to the `CoreKit` module, since we'll be using it later on.

AlexNaunt5 committed:
save data locally and improvement of the interface

AlexNaunt5 commented:
###  Allergy Management, Local Storage & Code Improvements


### **Enhancements Made:**

#### 1. **Allergy Input Handling**
- Replaced the single `allergies: String` with a dynamic `[String]` array.
  - Users can now **add multiple allergies individually**.
  - A list of added allergies is displayed in the form.
  - Each entry has a **"Remove" button** to easily delete mistakes.

#### 2. **Improved UX in the Form**
- Added a temporary input field (`allergyInput`) for individual allergy entry.
- Allergy section is clearly separated and interactive.
- Prevents adding empty or duplicate allergies.

#### 3. **Local Data Storage**
-  Implemented **local persistence** using `FileManager` and `JSONEncoder` / `JSONDecoder`.
  - Baby registration data is now saved to the app’s documents directory.
  - Data persists across app sessions, making it usable even after app relaunches.

#### 4. **ViewModel Logic Improvements**
- Added `addAllergy(_:)` and `removeAllergy(_:)` for better state management.
- Added `saveBaby()` and `loadBaby()` methods for encoding/decoding data.
- Improved validation logic remains in place.

---

### Next Steps / Suggestions
- Add screen confirmation on successful save.
- Optionally support editing/removing saved data from storage.
- Consider syncing with a remote repo or iCloud if needed.

GabrielVillagran commented on BabbleBuddyApp/BabbleBuddyApp/AppDelegate.swift:
```diff
@@ -6,12 +6,13 @@
 
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
-        // Override point for customization after application launch.
+        FirebaseApp.configure()
```
I recommend to upload just the changes made in your feature, if you upload all the files probably you or maybe all of us are going to face some problems when merge the PR, I'm telling you this because I faced this problem this past days. 

For instance if you are working on your DesignKit module, just upload everything related with that module, If it's necessary to make a call on the main bundle you can add it, but for instance, I think this is related with another feature you are working on and it's not necessary.

Regarding on how to run the app, what I usually do, is to make everything that is needed to run the app on the main app and upload just the files of the Feature, this problems (such as delete the storyboard reference, add the library and frameworks references) needs to be fixed on the main app ASAP in order to not doing the process every time we are going to run the app.

AlexNaunt5 replied:
The implementation of the Analytics features on BabyRegistrationForm was an error while changing of branch but the code generated for the Analytics was eliminated from the BabyRegistrationForm branch and now was moved to the corresponding features branch.

GabrielVillagran commented on CoreKit/AnalyticsManager.swift:
Same as before, this file is from another feature

AlexNaunt5 replied:
Code eliminated, it corresponds to the feature of Analytics and was moved to there.

GabrielVillagran commented on CoreKit/CoreKit.xcodeproj/project.pbxproj:
Same as before, this file is from another feature

AlexNaunt5 replied:
Code eliminated, it corresponds to the feature of Analytics and was moved to there.

GabrielVillagran commented on DesignKit/DesignKit/Baby.swift:
```diff
@@ -0,0 +1,38 @@
+    var height: Double
+    var heightUnit: HeightUnit
+    var bloodType: String
+    var allergies: [String]
```
This might work locally, but I’d recommend using the real model early on. That way, you won’t need to rework everything later when integrating with the actual service.

Also, since this is your data model, it would be great to align it with naming conventions such as BabyModel or BabyInformationModel.

PS: I wasn’t able to run the screen directly in the project and preview isn’t working for me. I'd suggest temporarily adding it to the main app bundle so you can test it in the simulator or even on a real device.

GabrielVillagran commented on DesignKit/DesignKit/BabyRegistrationViewModel.swift:
```diff
@@ -0,0 +1,99 @@
+
+/*
+func submitForm() {
+    AnalyticsManager.shared.logEvent("baby_registration_started")
```
Delete code that is not necessary, we should avoid comments on the code unless it is necessary to inform about something or probably a //MARK: (Text you want to add)

AlexNaunt5 replied:
Comment eliminated, it corresponds to the feature of Analytics and was moved to there.

GabrielVillagran commented on DesignKit/DesignKit/BabyRegistrationViewModel.swift:
```diff
@@ -0,0 +1,99 @@
+    @Published var showAlert = false
+    @Published var alertMessage = ""
+
+    func validateForm() -> Bool {
```
NIT: This function is doing a lot of comparisons, is there a way to implement a different logic?, I think you can create a function for each comparison and here you can just call them for verification

AlexNaunt5 replied:
Refactored the `validateForm()` method to improve readability and maintainability by extracting each validation into its own private method.

AlexNaunt5 committed:
delete elements of feature Crashlytics/Analytics (leave it for it's own branch)

AlexNaunt5 committed:
refactorize comparisions and data model

AlexNaunt5 commented:
Updated the data model and cleaned up unused features:

####  Changes made:
- Renamed `Baby` struct to `BabyModel` to follow naming conventions.
- Updated `BabyRegistrationViewModel` to use `BabyModel`.
- Implemented local data storage using `JSONEncoder`/`Decoder`.
- Removed unused Analytics-related code.

AlexNaunt5 committed:
eliminate reference to storyboard

DionicioCrz commented on BabbleBuddyApp/BabbleBuddyApp/AppDelegate.swift:
Instead of being doing this manually, why don't you point this branch to the branch you branched off?
And is a good practice to do rebasing to your pr with the develop to get the most recent changes.

AlexNaunt5 replied:
Thanks for the comment, I'll do next time

AlexNaunt5 committed:
add localizable strings file and pass the viewModel as a parameter instead of hardcoded

noelhpa reviewed (changes requested):
# This have the same changes of this #13  [PR](https://github.com/cincinnatiai/BabbleBuddyiOS/pull/5) this should not be done this way, if you are going to add changes for that branch you should create a branch form that and point the new changes to that branch
